File: solutions/python/pig-latin/1/pig_latin.py
def is_ascii_vowel(a: str) -> bool:
    return a in "aeiou"


def is_vowel_or_y(a: str) -> bool:
    return a in "aeiouy"


def starts_with_vowel(input: str) -> bool:
    return bool(input) and is_ascii_vowel(input[0])


def translate_word(input: str) -> str:
    # rule 1
    if starts_with_vowel(input) or input.startswith("xr") or input.startswith("yt"):
        return input + "ay"

    # rules 2 and 4
    consonants_count = 0
    if input.startswith("y"):
        consonants_count += 1
    while not is_vowel_or_y(input[consonants_count]):
        consonants_count += 1

    if consonants_count > 0:
        # rule 3
        if input[consonants_count - 1 : consonants_count + 1] == "qu":
            consonants_count += 1

        # back to rule 2
        return input[consonants_count:] + input[:consonants_count] + "ay"

    # fallback
    return input


def translate(text: str) -> str:
    return " ".join(translate_word(word) for word in text.split())

File: solutions/go/pig-latin/2/pig_latin.go
package piglatin

import (
	"slices"
	"strings"
)

var vowels = []rune("aeiou")

func is_vowel(c rune) bool {
	return slices.Contains(vowels, c)
}

func is_vowel_or_y(c rune) bool {
	return slices.Contains(vowels, c) || c == 'y'
}

func pig(input string) string {
	// rule 1
	prefix1 := rune(input[0])
	prefix2 := input[:2]
	if is_vowel(prefix1) || prefix2 == "xr" || prefix2 == "yt" {
		return input + "ay"
	}

	// rules 2 and 4
	consonants := 0
	if input[0] == 'y' {
		consonants++
	}
	for !is_vowel_or_y(rune(input[consonants])) {
		consonants++
	}

	if consonants > 0 {
		// rule 3
		if input[consonants-1:consonants+1] == "qu" {
			consonants++
		}

		// back to rule 2
		return input[consonants:] + input[:consonants] + "ay"
	}

	// fallback
	return input
}

func Sentence(sentence string) string {
	pig_words := []string{}
	for _, word := range strings.Fields(sentence) {
		pig_words = append(pig_words, pig(word))
	}
	return strings.Join(pig_words, " ")
}
